File: 11072.cpp
#include <bits/stdc++.h>

#define endl '\n'
#define pi pair<int, int>
#define vi vector<pair<int, int> >
#define lli long long int

using namespace std;

void swap(pi &p, pi &q);

lli distsq(pi p1, pi p2);

int orientation(pi p, pi q, pi r);

bool compare(pi p1, pi p2);

void convexhull(int n, vi &points, vi &S);

bool phi(pi p, vi &S, vi &points);

void imprimir(vi &algo){
    printf("[(%d, %d)", algo[0].first, algo[0].second);
    for(int i = 1; i < algo.size(); ++i)
        printf(", (%d, %d)", algo[i].first, algo[i].second);
    cout << "]" << endl;
}

pi p0;

int main(){
    int n, m;
    int x, y;
    while(cin >> n){
        vi points(n), S;
        for(int i = 0; i < n; ++i){
            cin >> x >> y;
            points[i] = make_pair(x, y);
        }
        convexhull(n, points, S);
        cin >> m;
        while(m--){
            cin >> x >> y;
            pi p = make_pair(x, y);
            if(phi(p, S, points)) cout << "inside" << endl;
            else cout << "outside" << endl;
        }
    }
    return 0;
}

lli distsq(pi p1, pi p2){
    lli x = (p2.first - p2.second), y = (p2.second - p1.second);
    return (x*x) + (y*y);
}

int orientation(pi p, pi q, pi r){
    int ans;
    lli val = ((q.second-p.second)*(r.first-q.first)) - ((r.second-q.second)*(q.first-p.first));
    if(!val) ans = 0;
    else ans = val > 0 ? 1:2;
    return ans;
}

bool compare(pi p1, pi p2){
    int alpha = orientation(p0, p1, p2);
    bool ans;
    if(alpha == 0) ans = distsq(p0, p2) >= distsq(p0, p1);
    else ans = alpha == 2 ? true:false;
    return ans;
}

void swap(pi &p, pi &q){
    p0 = q;
    q = p;
    p = p0;
}

void convexhull(int n, vi &points, vi &S){
    int min = 0, ymin = points[0].second;
    int x, y, m = 1;
    for(int i = 1; i < n; ++i){
        x = points[i].first; y = points[i].second;
        if((y < ymin) || (y==ymin && x < points[min].first)){
            ymin = y; min = i;
        }
    }
    swap(points[0], points[min]);
    sort(points.begin()+1, points.end(), compare);
    for(int i = 1; i < n; ++i){
        while(i < n-1 && (orientation(p0, points[i], points[i+1]) == 0)) ++i;
        points[m++] = points[i];
    }
    S.push_back(points[0]);
    S.push_back(points[1]);
    S.push_back(points[2]);
    for(int i = 3; i < m; ++i){
        int tam = S.size();
        while(tam > 1 && orientation(S[tam-2], S[tam-1], points[i]) != 2){
            S.pop_back(); --tam;
        }
        S.push_back(points[i]);
    }
    if(S[S.size()-1] == S[S.size()-2] || m < 3) S.pop_back();
}

bool phi(pi p, vi &S, vi &points){
    bool ans = true;
    for(int i = 1; i <= S.size() && ans; ++i){
        pi p1 = S[i-1], p2;
        if(i == S.size()) p2 = S[0];
        else p2 = S[i];
        int a = orientation(p1, p2, p);
        if(a == 1) ans = false;
        else if(a == 0){
            int xm = min(p1.first, p2.first), xM = max(p1.first, p2.first);
            int ym = min(p1.second, p2.second), yM = max(p1.second, p2.second);
            if(!(p.first >= xm && p.first <= xM && p.second >= ym && p.second <= yM))
                ans = false;
        }
    }
    return ans;
}

File: 11072.py
from sys import stdin
from functools import cmp_to_key

p0, points = None, None

def distsq(p1, p2):
	return ((p2[0]-p1[0])**2) + ((p2[1]-p1[1])**2)

def orientation(p, q, r):
	val, ans = ((q[1]-p[1])*(r[0]-q[0])) - ((r[1]-q[1])*(q[0]-p[0])), None
	if val == 0: ans = 0
	elif val > 0: ans = 1
	else: ans = 2
	return ans

def compare(p1, p2):
	alpha, ans = orientation(p0, p1, p2), None
	if alpha == 0:
		if distsq(p0, p2) >= distsq(p0, p1): ans = -1
		else: ans = 1
	else:
		if alpha == 2: ans = -1
		else: ans = 1
	return ans

def convexhull(n):
	global p0, points
	min, ymin = 0, points[0][1]
	for i in range(1, n):
		(x,y) = points[i]
		if (y < ymin) or (y==ymin and x < points[min][0]): min, ymin = i, y
	points[0], points[min] = points[min], points[0]
	p0, m = points[0], 1
	points.sort(key=cmp_to_key(compare))
	for i in range(1, n):
		while (i < n-1) and (orientation(p0, points[i], points[i+1])==0): i += 1
		points[m], m = points[i], m+1
	S = [points[i] for i in range(3)]
	for i in range(3, m):
		while (len(S) > 1) and (orientation(S[-2], S[-1], points[i]) != 2): S.pop()
		S.append(points[i])
	if S[-1] == S[-2]: S.pop()
	return S

def phi(p, S):
	ans, i = True, 1
	while ans and i <= len(S):
		p1, p2 = S[i-1], None 
		if i == len(S): p2 = S[0]
		else: p2 = S[i]
		a = orientation(p1, p2, p)
		if a == 1: ans = False
		elif a == 0:
			xm, xM = min(p1[0], p2[0]), max(p1[0], p2[0])
			ym, yM = min(p1[1], p2[1]), max(p1[1], p2[1]) 
			if not (p[0] >= xm and p[0] <= xM and p[1] >= ym and p[1] <= yM): ans = False
		i += 1
	return ans

def main():
	global points
	try:
		for temp in stdin:
			if temp.strip() == "": continue
			n = int(temp)
			points = [None for _ in range(n)]
			for i in range(n): points[i] = tuple(map(int, stdin.readline().split()))
			S = convexhull(n)
			for _ in range(int(stdin.readline().strip())):
				p = tuple(map(int, stdin.readline().split()))
				if phi(p, S): print("inside")
				else: print("outside")
			stdin.readline().strip()
	except:
		pass

main()